crates/red_knot_python_semantic/src/semantic_index/usedef.rs

```diff
@@ -0,0 +1,160 @@
+use crate::semantic_index::ast_ids::ScopedUseId;
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::ScopedSymbolId;
+use ruff_index::IndexVec;
+use std::ops::Range;
+
+/// All definitions that can reach a given use of a name.
+#[derive(Debug, PartialEq, Eq)]
+pub(crate) struct UseDefMap<'db> {
+    // TODO store constraints with definitions for type narrowing
+    all_definitions: Vec<Definition<'db>>,
```

Comment:
You could use an `IndexVec` here too, so that `Definitions::definitions` is a `Range<DefinitionId>`

Reply:
We can't use an `IndexVec` here because this has to list definitions in a specific order, and sometimes with the same definition listed more than once, in order to maintain that at any given point the active definitions for a symbol must all be adjacent in this vector. So it can't be just "all definitions ordered by definition ID". Hopefully this will be clearer in the follow-up PR with the module doc comment.